src/payment/strategies/bolt/bolt-payment-strategy.ts

```diff
@@ -62,10 +62,48 @@ export default class BoltPaymentStrategy implements PaymentStrategy {
         }
     }
 
-    finalize(): Promise<InternalCheckoutSelectors> {
-        return Promise.reject(new OrderFinalizationNotRequiredError());
+    /**
+     * The method that will be triggered by Bolt to finalize an order
+     * if 'Full Checkout with Fraud Protection' configuration mode enabled
+     *
+     * @param options PaymentRequestOptions
+     * @returns Promise<InternalCheckoutSelectors>
+     */
+    finalize(options?: PaymentFinalizeOptions): Promise<InternalCheckoutSelectors> {
+        if (this._useBoltClient) {
+            return Promise.reject(new OrderFinalizationNotRequiredError());
+        }
+
+        if (!options) {
+            throw new PaymentArgumentInvalidError(['options']);
+        }
+
+        const { methodId, paymentData } = options;
+
+        if (!methodId) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        if (!paymentData || !isNonceLike(paymentData)) {
+            throw new MissingDataError(MissingDataErrorType.MissingPayment);
+        }
+
+        const paymentPayload = {
+            methodId,
+            paymentData,
+        };
+
+        return this._store.dispatch(this._paymentActionCreator.submitPayment(paymentPayload));
```

Comment:
I'm not sure how it is even possible for you to submit payment without a valid payment token, which is only made available after submitting an order. If that's the case, you're asking the SDK client to call `checkoutService.submitOrder` and then `checkoutService.finalizeOrderIfNeeded` immediately afterwards. What you're effectively doing is changing the expectation that `finalizeOrderIfNeeded` can be called in isolation (i.e.: after redirecting from a hosted payment provider back to checkout). Although you're not technically making a non-additive change to the contract of `finalizeOrderIfNeeded`, you're in effect changing its observable behaviour therefore can be considered a breaking change.

Reply:
Bolt sometimes has some troubles on their side cause of the issues with orphan orders on their side and they don't want to let us submit an order before Bolt payment will pass. They will trigger `checkoutService.finalizeOrderIfNeeded` after Bolt payment will pass, that will trigger our submitPayment request

Reply:
@BC-TKozub can give to @davidchin some more information about the case?

Reply:
Sure. @davidchin, the main issue with current Bolt integration is submitting BC order after checkout is finished on Bolt side. (Bolt multi-step checkout). In some cases they have an issue with submitting BC order (connection issue, issues on shopper side, etc). As a result shopper has successfully completed checkout process but merchant does not have an order in control panel (orphan Bolt orders). The only way to fix current bihavior is do submit order via checkout-sdk before Bolt authorize payment and then finalize an order (perform submitPayment).

Reply:
Yes, I'm aware of the orphan order situation and I was the one who suggested to them to create BC order first. But what I was picturing is to do that as a part of `execute`, not `finalize`. If you do that, you wouldn't need to change the interface of `finalizeOrderIfNeeded`.

Reply:
Hi @davidchin! From bolt perspective, we need to have two different checkout SDK calls: one to create pre-order and the second to create a payment.
In the current PR proposed solution where we use **execute** for pre-order and and **finalize** for payment. If you think we need to split two actions in another way please let us know.

Reply:
Hi @vitaliyreznikov, I'm slightly confused. Is there any reason why you can't create the order and submit payment details (in that sequence) within `execute`? Also, we're the client of the Bolt library that they provided, they shouldn't care whether we have one method or two because they shouldn't have a reference to our Checkout SDK library.

Reply:
Hi @davidchin 

**1.** 
> Also, we're the client of the Bolt library that they provided
Yes and no. We have two different implementations in this file:
- _executeWithBoltClient: used in bolt product "fraud only", Checkout SDK calls Bolt library
- _executeWithBoltCheckout: used in bolt product "full checkout", bolt code calls Checkout SDK.

This PR changes _executeWithBoltCheckout only, so we are not going to add any cyclical dependency

**2. Why we need this**

Current flow for _executeWithBoltCheckout (full checkout):

```
1. Bolt creates transaction (bolt API call)
2. Bolt creates bigc order and payment (checkout SDK call)
```

desired flow:
```
1. Bolt creates bigc order (checkout SDK call)
2. Bolt creates transaction (bolt API call)
3. Bolt creates bigc payment (checkout SDK call)
```
In this flow, if (1) failed we are going to show an error instead of payment creation. This fixed the orphan order issue.

3. We can use another approach: instead of splitting checkout SDK call into two calls we can call bolt code in the middle. So after line 147 https://github.com/bigcommerce/checkout-sdk-js/blob/a46540bfcf50a313ebfe624dda59153f0ae1f983/src/payment/strategies/bolt/bolt-payment-strategy.ts#L147 add something like:

```
        const transactionReference = await this._boltClient.getTransactionReference();
        if (transactionReference === undefined) {
            return state;
        }
        paymentData.nonce = transactionReference;
```
It also solves issue because on bolt side we can actually start transaction only when we see getTransactionReference call.
Two advantages:
- code is very similar to existing code in _executeWithBoltClient implementation: https://github.com/bigcommerce/checkout-sdk-js/blob/a46540bfcf50a313ebfe624dda59153f0ae1f983/src/payment/strategies/bolt/bolt-payment-strategy.ts#L99
- no need to change finalizeOrderIfNeeded signature.

Please let me know if you think that this approach works better for you than the current PR.

Also if you have time I am happy to show you both implementation (_executeWithBoltClient and _executeWithBoltCheckout), how they look on client sites.

Reply:
Hello @davidchin, can you check last comment from @vitaliyreznikov, please?

Reply:
Thank you @vitaliyreznikov and @Sireks for coming up with a solution that doesn't involve changing the signature of `finalizeOrderIfNeeded`. It looks like a better option to me. 👍